process/dns_v2.go

```diff
@@ -0,0 +1,500 @@
+		bucketPosition := 0
+		if i > 0 {
+			bucketPosition = buckets[i-1].size
+		}
+
+		positionCounter += bucketPosition
+
+		positionBuffer = e.appendVarInt(positionBuffer, positionCounter)
+	}
+
+	return buffer, nil
+}
+
+func (e *V2DNSEncoder) EncodeDomainDatabase(names []string) ([]byte, []int32, error) {
+	if len(names) == 0 {
+		return nil, nil, nil
+	}
+	offsets := make([]int32, len(names))
+
+	// walk the list of strings, figure out how much size we need
+	bufferSize := e.varIntSize(len(names))
+	// compute the index of the middle buffer
+	indexOfMiddle := len(names) / 2
+	offsetOfMiddle := 0
+
+	for idx, val := range names {
+		if idx == indexOfMiddle {
+			offsetOfMiddle = bufferSize
+			bufferSize += e.varIntSize(offsetOfMiddle)
+			offsetOfMiddle = bufferSize
+		}
+
+		bufferSize += e.varIntSize(len(val))
+		bufferSize += len(val)
+
+	}
+	buffer := make([]byte, bufferSize)
+	metaBuffer := buffer[:0]
+	// write the number of names
+	metaBuffer = e.appendVarInt(metaBuffer, len(names))
+	// write the offset of the middle string
+	metaBuffer = e.appendVarInt(metaBuffer, offsetOfMiddle)
+
+	for idx, val := range names {
+		// need to store the offset of the beginning of each string, by index.
+		// when finally encoded, the consumers will get offsets into this
+		// buffer (for fast searching).
+		offsets[idx] = int32(len(metaBuffer))
+		metaBuffer = e.appendVarInt(metaBuffer, len(val))
+		metaBuffer = append(metaBuffer, val...)
+	}
+	return buffer, offsets, nil
+}
+
+func (e *V2DNSEncoder) varIntSize(value int) int {
+	return binary.PutUvarint(e.scratch[0:], uint64(value))
+}
+
+func (e *V2DNSEncoder) appendVarInt(buf []byte, value int) []byte {
+	bytesWritten := binary.PutUvarint(e.scratch[0:], uint64(value))
+
+	return append(buf, e.scratch[0:bytesWritten]...)
+}
+
+// getV2 returns a single offset into the name buffer for the first
+// domain string, followed by a slice of the offsets into the buffer
+// for the remaining strings.
+func getV2(buf []byte, ip string) (int32, []int32) {
+	var first int32 = -1
+	var names []int32
+
+	iterateDNSV2(buf, ip, func(i, total int, entry int32) bool {
+		if i == 0 {
+			first = entry
+			if total > 1 {
+				names = make([]int32, 0, total-1)
+			}
+		} else {
+			names = append(names, entry)
+		}
+		return true
+	})
+
+	return first, names
+}
+
+// returns a slice of all of the strings in the encodedDnsDomains list.
+func getDNSNameListV2(buf []byte) []string {
+	var names []string
+
+	num, bytesRead := binary.Uvarint(buf[0:])
+
+	// read the offset of the middle index; however, since we're reading
+	// the whole list we don't need it.
+	_, bytesReadForMiddle := binary.Uvarint(buf[bytesRead:])
+
+	bytesRead += int(bytesReadForMiddle)
+
+	for count := uint64(0); count < num && bytesRead < len(buf); count++ {
+		namelen, bytesReadForNameLen := binary.Uvarint(buf[bytesRead:])
+		bytesRead += bytesReadForNameLen
+		name := string(buf[bytesRead : bytesRead+int(namelen)])
```

Comment:
While running this code in prod, I ran into the following error:

```
panic: runtime error: slice bounds out of range [13:12]

goroutine 1633 [running]:
github.com/DataDog/dd-go/vendor/github.com/DataDog/agent-payload/process.getDNSNameListV2({0xc00744000b, 0xb6, 0x1fdff5})
	/go/src/github.com/DataDog/dd-go/vendor/github.com/DataDog/agent-payload/process/dns_v2.go:285 +0x246
github.com/DataDog/dd-go/vendor/github.com/DataDog/agent-payload/process.(*CollectorConnections).GetDNSNamesV2(...)
```

I will try to find what message is causing this. @derekwbrown, just sharing it here in case you can find some issue.

Reply:
Never mind, it was an issue with decoding.